shepherd

```diff
@@ -5,30 +5,101 @@ server_version() {
   docker version -f "{{.Server.Version}}"
 }
 
+# Runs `docker service inspect` with output formatting.
+# @param {string} service The service name
+# @param {string} format The format string
+service_inspect() {
+  local service="$1"
+  local format="$2"
+
+  docker service inspect "$service" -f "$format"
+}
+
 update_services() {
   local blacklist="$1"
   local supports_detach_option=$2
   local supports_registry_auth=$3
+  local manual_mode_for_all=$4
   local detach_option=""
   local registry_auth=""
+  local label_manual_mode="com.shepherd.manual_mode"
 
   [ $supports_detach_option = true ] && detach_option="--detach=false"
   [ $supports_registry_auth = true ] && registry_auth="--with-registry-auth"
 
   for service in $(IFS="\n" docker service ls --quiet); do
-    local name image_with_digest image
-    name="$(docker service inspect "$service" -f '{{.Spec.Name}}')"
+    local name service_image_with_digest service_image service_image_digest service_has_manual_mode
+    service_has_manual_mode=false
+
+    name="$(service_inspect "$service" '{{.Spec.Name}}')"
     if [[ " $blacklist " != *" $name "* ]]; then
-      image_with_digest="$(docker service inspect "$service" -f '{{.Spec.TaskTemplate.ContainerSpec.Image}}')"
-      image=$(echo "$image_with_digest" | cut -d@ -f1)
-      echo "Updating service $name with image $image"
-      docker service update "$service" $detach_option $registry_auth --image="$image" > /dev/null
+      # Format: [organization]/[repository]:[tag]@sha256:[checksum]
+      service_image_with_digest=$(service_inspect "$service" '{{.Spec.TaskTemplate.ContainerSpec.Image}}')
+      # Format: [organization]/[repository]:[tag]
+      service_image=$(echo "$service_image_with_digest" | cut -d@ -f1)
+      # Format: sha256:[checksum]
+      service_image_digest=$(echo "$service_image_with_digest" | cut -d@ -f2)
+
+      # Is "manual mode" enabled for all services via environment variable?
+      if [[ "$manual_mode_for_all" == true ]]; then service_has_manual_mode=true; fi
+      # ... otherwise, is "manual mode" enabled for this service via service label?
+      if [[ "$service_has_manual_mode" == false ]]; then
+        service_has_manual_mode=$(service_inspect "$service" '{{index .Spec.TaskTemplate.ContainerSpec.Labels "'"$label_manual_mode"'"}}')
+      fi
+
+      if [[ "$service_has_manual_mode" == true ]]; then
+        # We do not rely on Docker's image resolution and pulling logic to turn:
+        # `[organization]/[repository]:[tag]`
+        # ... into the latest: `[organization]/[repository]:[tag]@sha256:[checksum]`
+        # ... to then pull that image and update the service `$service` with it.
+        local local_image_with_digest local_image_digest
+
+        # This command is straightforward, it pulls the latest tagged image locally
+        # It also works with private repository images, if the correct volumes are
+        # mounted and `docker login` was executed on the host.
```

Comment:
Can you explain why (and which) volumes need to be mounted? A Docker pull does not need any..